docs/index.rst

```diff
@@ -69,6 +69,53 @@ Example:
    to this mount.
 
 
+CDN
+~~~
+
+Crane also supports some advanced options for v2 content that resides on a Content
+Delivery Network.
+
+The URL match and replace parameters support rewriting the URL where content is
+served on the CDN (redirect URL) to a new location. This can be used independently
+of the authorization features if desired.
+
+The URL authorization options allow Crane to support URL-based token authorization
+via Akamai's `Authorization Token 2.0 <https://community.akamai.com/people/B-3-2THQ57/blog/2016/02/26/ever-wondered-how-to-protect-your-assets-with-token-auth-edgeauth-20>`_
+specification. When authorization is used, Crane will append a query string parameter
+to all v2 redirects containing an expiration time and HMAC token. The field delimiter
+in the token is set to ``~`` based on Akamai's specificaton.
+
+Example URL with authorization token:
+
+::
+
+  https://cdn.example.com/path/to/content?_auth_=exp=1524660654~hmac=d039ac10e019fd13824a3f861b4f55df40e2a402d102b5266194fff6f3a24ed0
+
+The following options should go under a section named ``[cdn]``
+
+url_match
+  optional portion of URL to match when rewriting content redirect URLs
+
+url_replace
+  optional replacement for ``url_match`` when rewriting the redirect URL. This will
+  only be used when specified along with ``url_match``
+
+url_auth_secret
```

Comment:
since it crucial to provide a hexadecimal string we need to validate the in the config parsing, otherwise we get an ugly 500
```
[root@ina crane]# curl  -H "Accept: application/vnd.docker.distribution.manifest.v2+json" -X GET -k 'http://localhost:1234/v2/ina/manifests/latests'
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>500 Internal Server Error</title>
</head><body>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error or
misconfiguration and was unable to complete
your request.</p>
<p>Please contact the server administrator at 
 root@localhost to inform them of the time this error occurred,
 and the actions you performed just before this error.</p>
<p>More information about this error may be available
in the server error log.</p>
</body></html>
[root@ina crane]# less /var/log/httpd/error_log| tail -30
[Thu Apr 26 12:52:54.038862 2018] [wsgi:error] [pid 6178] [INFO] crane.app: application initialized
[Thu Apr 26 12:52:54.039988 2018] [wsgi:error] [pid 6178] [remote ::1:120] mod_wsgi (pid=6178): Exception occurred processing WSGI script '/usr/share/crane/crane.wsgi'.
[Thu Apr 26 12:52:54.040118 2018] [wsgi:error] [pid 6178] [remote ::1:120] Traceback (most recent call last):
[Thu Apr 26 12:52:54.040179 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/usr/lib64/python2.7/site-packages/flask/app.py", line 2000, in __call__
[Thu Apr 26 12:52:54.040824 2018] [wsgi:error] [pid 6178] [remote ::1:120]     return self.wsgi_app(environ, start_response)
[Thu Apr 26 12:52:54.040861 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/usr/lib64/python2.7/site-packages/flask/app.py", line 1991, in wsgi_app
[Thu Apr 26 12:52:54.040884 2018] [wsgi:error] [pid 6178] [remote ::1:120]     response = self.make_response(self.handle_exception(e))
[Thu Apr 26 12:52:54.040919 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/usr/lib64/python2.7/site-packages/flask/app.py", line 1567, in handle_exception
[Thu Apr 26 12:52:54.040941 2018] [wsgi:error] [pid 6178] [remote ::1:120]     reraise(exc_type, exc_value, tb)
[Thu Apr 26 12:52:54.040977 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/usr/lib64/python2.7/site-packages/flask/app.py", line 1988, in wsgi_app
[Thu Apr 26 12:52:54.040998 2018] [wsgi:error] [pid 6178] [remote ::1:120]     response = self.full_dispatch_request()
[Thu Apr 26 12:52:54.041029 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/usr/lib64/python2.7/site-packages/flask/app.py", line 1641, in full_dispatch_request
[Thu Apr 26 12:52:54.041049 2018] [wsgi:error] [pid 6178] [remote ::1:120]     rv = self.handle_user_exception(e)
[Thu Apr 26 12:52:54.041085 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/usr/lib64/python2.7/site-packages/flask/app.py", line 1544, in handle_user_exception
[Thu Apr 26 12:52:54.041105 2018] [wsgi:error] [pid 6178] [remote ::1:120]     reraise(exc_type, exc_value, tb)
[Thu Apr 26 12:52:54.041140 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/usr/lib64/python2.7/site-packages/flask/app.py", line 1639, in full_dispatch_request
[Thu Apr 26 12:52:54.041160 2018] [wsgi:error] [pid 6178] [remote ::1:120]     rv = self.dispatch_request()
[Thu Apr 26 12:52:54.041179 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/usr/lib64/python2.7/site-packages/flask/app.py", line 1625, in dispatch_request
[Thu Apr 26 12:52:54.041221 2018] [wsgi:error] [pid 6178] [remote ::1:120]     return self.view_functions[rule.endpoint](**req.view_args)
[Thu Apr 26 12:52:54.041231 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/home/ipanova/pulp_development/crane/crane/views/v2.py", line 125, in name_redirect
[Thu Apr 26 12:52:54.041352 2018] [wsgi:error] [pid 6178] [remote ::1:120]     auth_token = app_util.generate_cdn_url_token(cdn_path, auth_secret, auth_exp)
[Thu Apr 26 12:52:54.041371 2018] [wsgi:error] [pid 6178] [remote ::1:120]   File "/home/ipanova/pulp_development/crane/crane/app_util.py", line 340, in generate_cdn_url_token
[Thu Apr 26 12:52:54.041502 2018] [wsgi:error] [pid 6178] [remote ::1:120]     binascii.a2b_hex(secret),
[Thu Apr 26 12:52:54.041535 2018] [wsgi:error] [pid 6178] [remote ::1:120] TypeError: Odd-length string
[Thu Apr 26 12:52:54.113171 2018] [wsgi:error] [pid 6178] [INFO] crane.data: loading metadata from /var/lib/pulp/published/docker